torch/overrides.py

```diff
@@ -1511,10 +1511,10 @@ def handle_torch_function(
 
     # Check for __torch_function__ mode.
     if _is_torch_function_mode_enabled():
-        mode = _get_current_function_mode()
-        # NB: unlike on tensors, modes are instances
-        with _no_torch_function_mode():
-            result = mode.__torch_function__(public_api, types, args, kwargs)
+        # if we're here, the mode must be set to a TorchFunctionStackMode
+        # this unsets it and calls directly into TorchFunctionStackMode's torch function
+        _set_torch_function_mode(None)
+        result = _TorchFunctionStackMode().__torch_function__(public_api, types, args, kwargs)
```

Comment:
The code smell here, is that the unsetting of the torch function mode is not scoped, so you're destroying whatever the previous state was by the time you come out. Maybe that's what the old code is, but it's not obvious to me that this is right either.